Allowed non-this, non-super code before super call in derived classes with property initializers

JoshuaKGoldberg opened the pull request:
Starts on #8277 by allowing the non-`this`, non-`super` code to be root-level statements in the constructor. This will now be allowed:

```ts
class Base { }
class Derived extends Base {
    public prop = true;
    constructor(public paramProp = true) {
        console.log("Hello, world!");
        super();
    }
}
```

~It feels wrong to put a new `forEachChild` loop in the checker, though in the vast majority of user files this will be a very quick one. Is there a better way to check for a reference to `super` or `this`?~

**Edit 2/28/2020**: I've ~mostly~ resolved the merge conflicts introduced by both `#` private fields & `useDefineForClassFields` ~, but I'm not confident my approach is still a valid one. I'd greatly appreciate it if someone could confirm I'm on the right track~!
~Oh, and `gulp runtests` passes locally (on Windows). I'll try a Mac to see if there's some odd encoding/whitespace behavior with the failing test...~ ✔️ 

**Edit 3/11/2021**: It seems this is fairly close to merging. 

**Edit 1/13/2022**: I bought this PR a birthday cake for its third birthday. https://twitter.com/JoshuaKGoldberg/status/1481654056422567944

<img alt="Blue and white cake with text 'Happy 3rd birthday, #29374 Bump for PR review please!' and the TypeScript logo. A tall slice is removed." src="https://user-images.githubusercontent.com/3335181/149387913-4756dcb0-dcb4-49c9-a9b1-f4729e6bda68.jpeg" width="490px" />

ajafff commented on src/compiler/checker.ts:
```diff
@@ -23628,6 +23628,20 @@ namespace ts {
+                    return true;
+                }
+
+                if (isClassLike(node) || isFunctionLike(node)) {
```
isFunctionLike includes immediately executed arrow functions or object literal methods/accessors with computed names, that could use `this`

JoshuaKGoldberg replied:
Added a new `isNewThisScope` method - I think that should cover it?

weswigham commented:
> It feels wrong to put a new forEachChild loop in the checker, though in the vast majority of user files this will be a very quick one. Is there a better way to check for a reference to super or this?

`TransformFlags.Super | TransformFlags.ContainsSuper` for `super` at least.

ajafff commented on src/compiler/utilities.ts:
```diff
@@ -6258,6 +6258,29 @@ namespace ts {
+    export function isNewThisScope(node: Node): boolean {
+        switch (node.kind) {
+            case SyntaxKind.ClassDeclaration:
+            case SyntaxKind.ClassExpression:
```
`class C extends this.OtherClass {}` the `this` actually refers to the containing object or class.
The same applies to computed names of methods and accessors in that class (although TypeScript currently doesn't allow that)

JoshuaKGoldberg commented on tests/baselines/reference/derivedClassSuperProperties.errors.txt:
```diff
@@ -0,0 +1,363 @@
+    ~~~~~~~~~~~~~~~~
+        }
+    ~~~~~
+!!! error TS2376: A 'super' call must be the first statement in the constructor to refer to 'super' or 'this' when a derived class contains initialized properties or has parameter properties.
```
☹ this test case shouldn't be failing. Will fix.

JoshuaKGoldberg replied:
Fixed in 00b4a2bd2a74bae49305e988d7315d507cbb7b66

ajafff commented on src/compiler/utilities.ts:
```diff
@@ -6258,6 +6258,34 @@ namespace ts {
+                        return false;
+                }
+            case SyntaxKind.FunctionDeclaration:
+            case SyntaxKind.FunctionExpression:
```
Now that you removed classes, you need to add `ConstructorDeclaration` and `PropertyDeclaration`.
The latter one makes the `case SyntaxKind.ArrowFunction` redundant.

ajafff commented on src/compiler/checker.ts:
```diff
@@ -23628,6 +23628,18 @@ namespace ts {
             }
         }
 
+        function nodeRefersToSuperOrThis(node: Statement): boolean {
```
Consider renaming `refersTo` to `references` to make it more consistent with similar function names

ajafff commented on src/compiler/utilities.ts:
```diff
@@ -6258,6 +6258,34 @@ namespace ts {
     }
 
+    /* @internal */
+    export function isNewThisScope(node: Node): boolean {
```
It's probably worth describing what this function does. Maybe you could even inline it.

ajafff commented on src/compiler/utilities.ts:
```diff
@@ -6258,6 +6258,34 @@ namespace ts {
+                switch (node.parent.kind) {
+                    case SyntaxKind.GetAccessor:
+                    case SyntaxKind.MethodDeclaration:
+                    case SyntaxKind.SetAccessor:
```
Consider adding comments that describe why these nodes have special handling.

ajafff commented on src/compiler/utilities.ts:
```diff
@@ -6258,6 +6258,34 @@ namespace ts {
+                }
+            case SyntaxKind.FunctionDeclaration:
+            case SyntaxKind.FunctionExpression:
+            case SyntaxKind.Parameter:
```
this will falsely abort recursing into parameters of arrow functions

JoshuaKGoldberg replied:
Added a test case: `this` cannot be referenced in constructor arguments is the actual expected error there... which seems odd. I'll look into why and might file a separate issue.

ajafff replied:
I created a PR to fix that error this week, don't know if it's already merged

JoshuaKGoldberg replied:
For posterity: https://github.com/Microsoft/TypeScript/pull/29298

ajafff commented on tests/baselines/reference/derivedClassSuperProperties.js:
```diff
@@ -0,0 +1,562 @@
+var DerivedWithFunctionDeclaration = /** @class */ (function (_super) {
+    __extends(DerivedWithFunctionDeclaration, _super);
+    function DerivedWithFunctionDeclaration() {
+        var _this = this;
```
Transpiled property initializers/parameter properties still access `this` before calling `super()`. There's probably a condition in the TS transform that checks if the first executable statement is `super();`. That's no longer correct

JoshuaKGoldberg replied:
Sorry, I'm not sure what change you'd like to see here?

This is already an allowed edge case:

```typescript
class Base { }

class Child extends Base {
    constructor() {
        (() => this)();
        super();
    }
}

new Child()
```

TypeScript gives no checker complaint, and its downpiled output runs without error, but running that directly gives the "Must call super constructor..." error.

ajafff replied:
I don't mean changing the error behavior here. Instead I'm pointing out that the transpiled code is wrong. That's more obvious if you transpile to ES6

```ts
class C extends Object {
  prop = 1;
  constructor() {
    something();
    super();
  }
}

// currently transpiles to something similar to this
class C extends Object {
  constructor() {
    this.prop = 1;  // this gives an error when executed
    return super() || this;
  }
}

// instead it should be
class C extends Object {
  constructor() {
    const _this = super() || this;
    this.prop = 1;
    return _this;
  }
}
```

JoshuaKGoldberg replied:
Gotcha, thanks!

This looks to be in `src/compiler/transformers/es2015.ts` within `declareOrCaptureOrReturnThisForConstructorIfNeeded`.

JoshuaKGoldberg commented on tests/baselines/reference/derivedClassSuperProperties.errors.txt:
```diff
@@ -0,0 +1,425 @@
+    ~~~~~~~~~~~~~~~~
+        }
+    ~~~~~
+!!! error TS2376: A 'super' call must be the first statement in the constructor to refer to 'super' or 'this' when a derived class contains initialized properties or has parameter properties.
```
🤔 ...I'm thinking this should be allowed, no? It's a variation of:

```ts
class DerivedWithArrowFunction extends Base {
    prop = () => this;

    constructor() {
        const arrow = () => this;
        super();
    }
}
```

ajafff commented on src/compiler/utilities.ts:
```diff
@@ -6258,6 +6258,33 @@ namespace ts {
+     * @internal
+     */
+    export function isNewThisScope(node: Node): boolean {
+        switch (node.kind) {
```
all kinds of decorators always refer to the scope containing the class declaration:

```ts
@decorator(this)
class C {
  @decorator(this)
  prop = 1;
  @decorator(this)
  method(@decorator(this) foo: string) {}
}
```

JoshuaKGoldberg replied:
Added test cases 👍

JoshuaKGoldberg commented:
Ping @ahejlsberg - is there anything that needs to be done here? It'd be nice to have this in 😄

JoshuaKGoldberg commented:
Correction: ping, @weswigham?

weswigham commented on src/compiler/checker.ts:
```diff
@@ -25724,24 +25724,33 @@ namespace ts {
-                        let superCallStatement: ExpressionStatement | undefined;
+                        // Until we have better flow analysis, it is an error to place the super call within any kind of block or conditional
+                        // See GH #8277
+                        if (superCall.parent.parent !== node.body) {
```
`(super())`?

JoshuaKGoldberg replied:
:+1: Added parenthesis diving with tests for `(super());` alone, before a `this.prop;`, and after a `this.prop;`.

rbuckton commented on tests/baselines/reference/derivedClassSuperProperties.js:
```diff
@@ -0,0 +1,844 @@
+var DerivedWithFunctionDeclaration = /** @class */ (function (_super) {
+    __extends(DerivedWithFunctionDeclaration, _super);
+    function DerivedWithFunctionDeclaration() {
+        var _this = this;
```
The emit here has the wrong runtime semantics.

rbuckton commented on tests/baselines/reference/derivedClassSuperProperties.js:
```diff
@@ -0,0 +1,844 @@
+    __extends(DerivedWithFunctionDeclarationAndThisParam, _super);
+    function DerivedWithFunctionDeclarationAndThisParam() {
+        var _this = this;
+        _this.prop = true;
```
The emit here has the wrong runtime semantics.

rbuckton commented on tests/baselines/reference/derivedClassSuperProperties.js:
```diff
@@ -0,0 +1,844 @@
+    __extends(DerivedWithFunctionExpression, _super);
+    function DerivedWithFunctionExpression() {
+        var _this = this;
+        _this.prop = true;
```
The emit here has the wrong runtime semantics.

rbuckton commented on tests/baselines/reference/derivedClassSuperProperties.js:
```diff
@@ -0,0 +1,844 @@
+    __extends(DerivedWithParenthesis, _super);
+    function DerivedWithParenthesis() {
+        var _this = this;
+        _this.prop = true;
```
The emit here has the wrong runtime semantics.

rbuckton replied:
The emit here is still incorrect. The `_this.prop = true;` statement needs to come after `(_this = _super.call(this) || this);`.

rbuckton commented on tests/baselines/reference/derivedClassSuperProperties.js:
```diff
@@ -0,0 +1,844 @@
+    __extends(DerivedWithParenthesisBeforeStatement, _super);
+    function DerivedWithParenthesisBeforeStatement() {
+        var _this = this;
+        _this.prop = true;
```
The emit here has the wrong runtime semantics.

rbuckton replied:
This is still incorrect.

rbuckton commented on tests/baselines/reference/derivedClassSuperProperties.js:
```diff
@@ -0,0 +1,844 @@
+    __extends(DerivedWithClassDeclaration, _super);
+    function DerivedWithClassDeclaration() {
+        var _this = this;
+        _this.prop = true;
```
The emit here has the wrong runtime semantics.

JoshuaKGoldberg commented on tests/baselines/reference/derivedClassSuperProperties.js:
```diff
@@ -0,0 +1,844 @@
+    __extends(DerivedWithArrowFunctionParameter, _super);
+    function DerivedWithArrowFunctionParameter() {
+        var _this = this;
+        _this.prop = true;
```
Thanks @rbuckton, this explanation is really helpful. I _think_ the transformers are fixed in afc58e123b2ab53627ec1721b1849ff8507af860, by having the emitted constructors contain, in order:

1. Pre-`super()` code
2. The `super()` itself
3. Properties with initializers
4. Parameter properties
5. Post-`super()` code

The `_a = () => {` in those snippets have me stumped, though. Is there a reason the contents can't be placed where the `_a();` is?

The updated baselines seem generally "better" than before. Meaning: wWithin the existing test cases that result in semantic errors for `this` access before `super`, several previously resulted in a global `_this` being created; now, they correctly defer referring to `_this` until it's assigned to `_super.call(this) || this`. Is that... good?

JoshuaKGoldberg replied:
Ping @rbuckton - do you think this could make it in for 3.8?

rbuckton replied:
> The `_a = () => {` in those snippets have me stumped, though. Is there a reason the contents can't be placed where the `_a();` is?

I think [my comment below](https://github.com/microsoft/TypeScript/pull/29374#discussion_r478785677) might provide a better example. ES2015 constructors allow you to call `super()` anywhere in the constructor, such as like this:

```ts
class Derived extends Base {
  prop = 0;
  constructor(f) {
    if (f) {
      super(1);
    }
    else {
      console.log(super(2));
    }
  }
  toString() { return this.prop.toString(); }
}
```

If you were to inline the property assignment, you'd end up with duplication and even a runtime error in the `else` case above:

```js
// @target: es2015
class Derived extends Base {
  constructor(f) {
    if (f) {
      super(1);
      this.prop = 0; // duplication
    }
    else {
      console.log(super(2)); // error because `this.prop` is undefined when `toString()` is called.
      this.prop = 0; // duplication
    }
   toString() { return this.prop.toString(); }
  }
```

By wrapping the `super` call in an arrow, you can preserve the correct evaluation order (I'll use `_super_1` instead of `_a` for clarity below):

```js
// @target: es2015
class Derived extends Base {
  constructor(f) {
    let _super_1 = (...args) => {
      super(...args);
      this.prop = 0;
      return this;
    };
   
    if (f) {
      _super_1(1);
    }
    else {
      console.log(_super_1(2)); // no error because `this.prop` is defined.
    }
  }
  toString() { return this.prop.toString(); }
}
```

Obviously, downlevel to ES5 is a bit more verbose:
```js
// @target: es5
var Derived = /** @class */ (function (_super) {
    __extends(Derived, _super);
    function Derived(f) {
        var _this = this;
        var _super_1 = function () {
            _this = _super.apply(this, arguments) || this;
            _this.prop = true;
            return _this;
        };

        if (f) {
          _super_1(1);
        }
        else {
          console.log(_super_1(2));
        }
    }
    Derived.prototype.toString = function () { return this.prop.toString(); };
}(Base));
```

The question is whether we want to change the emit in this PR, or introduce a separate PR that allows `super()` calls anywhere in the constructor. I'm tempted to ask that we do this now, even for this case, so that the emit for existing code doesn't end up going through multiple changes as we release new versions.

@DanielRosenwasser do you have any thoughts here?

JoshuaKGoldberg commented on tests/baselines/reference/strictModeInConstructor.js:
```diff
@@ -102,10 +103,10 @@ var C = /** @class */ (function (_super) {
     function D() {
-        var _this = this;
+        var x = 1; // No error
+        var y = _this.s; // Error
```
This is interesting. Because there is no longer a `var _this = this;` added to the beginning of the constructor, a reference to `_this` before `super()` in a derived constructor will now result in a `ReferenceError: _this is not defined`.

```ts
constructor() {
    this.weirdError(); // ReferenceError: _this is not defined
    super();
}
```

I'll look into fixing..

JoshuaKGoldberg commented:
Well this is fun: private named instance fields (#30829) did some merge conflict damage here 😅. I'll merge from master best I can now with a `gulp baseline-accept` to get this cleaned up, attempt to get the tests reasonable, and start pinging folks in earnest this week.

JoshuaKGoldberg commented on tests/baselines/reference/classUpdateTests.js:
```diff
@@ -190,7 +190,7 @@ var G = /** @class */ (function (_super) {
 var H = /** @class */ (function () {
     function H() {
-        _this = _super.call(this) || this;
+        return _super.call(this) || this;
```
Note that this is a behavioral change for when `H` is used incorrectly as a callable function:

```
let invalid = H();
```

...but the new output is a correction, as when `H` is changed to `extend Object`,  the `return` does exist.

JoshuaKGoldberg commented on tests/baselines/reference/defineProperty(target=es5).js:
```diff
@@ -42,6 +42,7 @@ var _a;
 var x = "p";
 var A = /** @class */ (function () {
     function A(y) {
+        this.y = y;
```
😡

JoshuaKGoldberg commented on src/compiler/checker.ts:
```diff
@@ -1090,8 +1090,8 @@ namespace ts {
-                     isAssignmentDeclaration(target.valueDeclaration) && !isAssignmentDeclaration(source.valueDeclaration) ||
-                     isEffectiveModuleDeclaration(target.valueDeclaration) && !isEffectiveModuleDeclaration(source.valueDeclaration))) {
+                        isAssignmentDeclaration(target.valueDeclaration) && !isAssignmentDeclaration(source.valueDeclaration) ||
+                        isEffectiveModuleDeclaration(target.valueDeclaration) && !isEffectiveModuleDeclaration(source.valueDeclaration))) {
```
I'll remove these whitespace changes.

JoshuaKGoldberg commented on src/compiler/transformers/classFields.ts:
```diff
@@ -663,23 +677,48 @@ namespace ts {
+                    }
                 }
-                if (afterParameterProperties > indexOfFirstStatement) {
-                    if (!useDefineForClassFields) {
```
This is a likely culprit for something removed in merge conflict resolution...

JoshuaKGoldberg commented:
@weswigham I believe I've addressed all the discussions on this PR, and it should be ready for re-review when you have a chance please!

weswigham commented on src/compiler/utilities.ts:
```diff
@@ -2898,6 +2898,33 @@ namespace ts {
+                switch (node.parent.kind) {
+                    case SyntaxKind.Constructor:
+                    case SyntaxKind.MethodDeclaration:
+                    // The `extends` clause of a class is its parent scope, unlike its constructor and methods
```
Is this comment still relevant?

JoshuaKGoldberg commented:
Ping @rbuckton - now that 3.9 is out, I think it'd be lovely to get this into 4.0. Do you know if you'll be able to review this? (please 🙏 )

DanielRosenwasser commented:
Was chatting with @ahejlsberg about this about 2 weeks ago - I recall agreeing that it'd be useful for 4.0, short of maybe some style nits. As a related change, he was also looking into ]our `this`-before-`super()` checks being too rudimentary. Maybe he can give an update on the whole picture.

btw, one thing I'd recommend is reviewing with whitespace off (https://github.com/microsoft/TypeScript/pull/29374/files?w=1)

rbuckton commented on tests/baselines/reference/derivedClassSuperStatementPosition.errors.txt:
```diff
@@ -0,0 +1,63 @@
+        prop = 1;
+        constructor() {
+            Math.random()
+                ? super(1)
```
Perhaps for a future PR, but it would be nice to actually support this. Its technically possible:

```ts
// ts
class Derived extends Base {
  prop = 1;
  constructor() {
    Math.random() ? super(1) : super(0);
  }
}
// es2015
class Derived extends Base {
  constructor() {
    // wrap `super` and initialization in an arrow
    const _super = (...args) => {
      super(...args);
      this.prop = 1;
      return this;
    };

    // class body
    Math.random() ? _super(1) : _super(0);
  }
}
// es5
var Derived = /** @class */ (function (_super) {
  __extends(Derived, _super);
  function Derived() {
    var _this = this;
    // wrap `super` and initialization in a function expression
    var _super_1 = function() {
      _this = _super.apply(_this, arguments) || _this;
      _this.prop = 1;
      return _this;
    };

    // class body
    Math.random() ? _super_1(1) : _super_1(0);
  }
  return Derived;
}(Base));
```

rbuckton commented on src/compiler/checker.ts:
```diff
@@ -29277,6 +29285,28 @@ namespace ts {
+        function superCallIsRootLevelInConstructor(superCall: Node, body: Block) {
+            let expressionParent = superCall.parent;
+
+            while (isParenthesizedExpression(expressionParent)) {
```
We already have a function that does this, called `walkUpParenthesizedExpressions`.

rbuckton commented on src/compiler/checker.ts:
```diff
@@ -29277,6 +29285,28 @@ namespace ts {
+                expressionParent = expressionParent.parent;
+            }
+
+            return expressionParent.parent === body;
```
Depending on what you mean by "root level", this check may be inadequate. Consider the following cases, which would return `true` for this function if they were in the constructor body:

```ts
for(super();;) {}
for(a; super();) {} // other arbitrary code can execute in 'a'
for(a; b; super()) {} // other arbitrary code can execute in 'a' and 'b'
for(; ; super()) { break; } // `super()` here will never be executed.
for (const x of super()) {}
while (super()) {}
do {} while (super());
if (super()) {}
switch (super()) {}
```

If the above cases are expected, I'd like to see tests that illustrate the intended behavior.

JoshuaKGoldberg commented on src/compiler/utilities.ts:
```diff
@@ -2898,6 +2898,31 @@ namespace ts {
+                    case SyntaxKind.MethodDeclaration:
+                    case SyntaxKind.GetAccessor:
+                    case SyntaxKind.SetAccessor:
+                        // Object properties can have computed names; only method-like bodies start a new scope
```
I don't believe we do? In the above, the `() => this.x` is the new scope, which should be covered by `SyntaxKind.ArrowFunction`.

JoshuaKGoldberg commented on src/compiler/transformers/es2015.ts:
```diff
@@ -1064,19 +1074,60 @@ namespace ts {
+                        ]
                     ),
-                    /*location*/ constructor.body.statements
+                    /*location*/ bodyStatements
```
Thanks, I think the mapping of text ranges being from the original node(s) to the new factory-generated ones just clicked for me...

JoshuaKGoldberg commented on tests/baselines/reference/emitSuperCallBeforeEmitParameterPropertyDeclaration1.js:
```diff
@@ -36,9 +36,10 @@ var A = /** @class */ (function () {
 var B = /** @class */ (function (_super) {
     __extends(B, _super);
     function B(x) {
+        var _this = this;
```
Figured this out after a bit of digging: `splitConstructorBodyStatementTypes` was putting all of original constructor body statements other than the super call in `bodyStatements` or `preSuperStatements`, which moved existing prologue statements into the list of ones added after the new `prologue` array. 9022a17 adds a returned `existingPrologue` array to directly capture prologue statements.

JoshuaKGoldberg commented on tests/baselines/reference/privateNameFieldAssignment.js:
```diff
@@ -52,8 +52,8 @@ var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (
     constructor() {
-        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m;
         _field.set(this, 0);
+        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m;
```
I bet it's from the prologue ordering being more manual (https://github.com/microsoft/TypeScript/pull/29374#discussion_r533913684). Do you see it as an issue? I don't see a case where this would cause a real behavioral difference in the emitted code.

JoshuaKGoldberg commented on tests/baselines/reference/parameterInitializerBeforeDestructuringEmit.js:
```diff
@@ -41,6 +41,8 @@ function foobar(_a) {
 }
 var C = /** @class */ (function () {
     function C(_a) {
+        "use strict";
+        "Some other prologue";
```
Oy, I re-requested review a little early... looking at this now.

JoshuaKGoldberg commented:
Ping @rbuckton - is there anything else you'd like changed?

> This PR doesn't have any linked issues. Please open an issue that references this PR. From there we can discuss and prioritise.

cc @DanielRosenwasser; I'm hesitant to mark this as _fixing_ the linked issue, as it only partially does that. Edge case in the tooling maybe?

andrewbranch commented:
Might be good to run the perf tests for good measure after merge conflict is fixed

rbuckton commented:
We discussed this in the design meeting today and look forward to this change. I hope to conclude my review shortly. In the meantime, could you address the merge conflict with _diagnosticMessages.json_?

rbuckton commented on src/compiler/transformers/ts.ts:
```diff
@@ -1943,13 +1943,19 @@ namespace ts {
             resumeLexicalEnvironment();
 
-            indexOfFirstStatement = addPrologueDirectivesAndInitialSuperCall(factory, constructor, statements, visitor);
+            const { foundSuperStatement, indexOfFirstStatementAfterSuper, indexAfterLastPrologueStatement } = addPrologueDirectivesAndInitialSuperCall(factory, constructor, statements, visitor);
```
As I stated below for `addPrologueDirectivesAndInitialSuperCall`, I think it would be better if we just broke this down into the component steps and inlined them here:

1. Copy/transform standard prologue directives to the new statement body using `factory.copyPrologue`
2. Find the index of the first `super` statement, and copy/transform statements between the prologue and the `super` call.
3. Transform the `super` call
4. Transform initializers and parameter-property assignments.
5. Transform the remaining statements.

That way we don't have to jump around within `statements` and prefix some statements and postifx others like we're doing on lines 1951-1956 and 1973-1980 as its just becoming to difficult to reason over.

rbuckton commented on src/compiler/transformers/es2015.ts:
```diff
@@ -1089,19 +1099,77 @@ namespace ts {
+        }
 
-            return block;
+        function splitConstructorBodyStatementTypes(originalBodyStatements: NodeArray<Statement>) {
```
I think this is introducing too much complexity. We probably need to inline this into `splitConstructorBodyStatementTypes` and find ways of simplifying the behavior. I've added a similar comment in _ts.ts_.

rbuckton commented on src/compiler/transformers/es2015.ts:
```diff
@@ -980,40 +980,40 @@ namespace ts {
             const statements: Statement[] = [];
             resumeLexicalEnvironment();
 
+            // In derived classes, there may be code before the necessary super() call
```
Part of the existing complexity here is my fault. The goal was to produce a statement list that contained the following statements in a specific order:

1. Existing *standard* prologue directives (i.e., `"use strict";`) from the original statement list.
2. New *custom* prologue statements (not really prologue directives, really just code injected into the top of the function body)
3. Existing *custom* prologue statements (again, not true prologue directives, but code injection used by transformations.
4. (new) Statements that come before `super()`
5. Injected code for `this` capturing as it applies to `super()` invocation.
6. The `super()` invocation itself.
7. Any remaining statements.

Also, the `mergeLexicalEnvironment` call injects temporary variable declarations between (2) and (3) above.

rbuckton commented on src/compiler/transformers/classFields.ts:
```diff
@@ -664,23 +678,57 @@ namespace ts {
-            addPropertyStatements(statements, properties, factory.createThis());
 
-            // Add existing statements, skipping the initial super call.
+            addPropertyStatements(statements, properties, factory.createThis(), getPropertyStatementInsertionIndex(needsSyntheticConstructor, foundSuperStatement, useDefineForClassFields ? 0 : parameterPropertyDeclarationCount));
```
Does `getPropertyStatementInsertionIndex` need to be its own function? It seems a conditional would have served just as well:

```suggestion
            const insertionIndex = needsSyntheticConstructor ? 1 : 
                foundSuperStatement ? undefined :
                useDefineForClassFields ? 0 :
                parameterPropertyDeclarationCount;
            addPropertyStatements(statements, properties, factory.createThis(), insertionIndex));
```

Then again, it'd be better to find a way to do this that doesn't use `splice`

rbuckton commented on src/compiler/transformers/classFields.ts:
```diff
@@ -664,23 +678,57 @@ namespace ts {
+                        }
+                        // Since there wasn't a super() call, add them to the top of the constructor
+                        else {
+                            statements = addRange(
```
Why are you using `addRange` with an empty array here? Why not just do `[...parameterProperties, ...statements]`?

rbuckton commented on src/compiler/transformers/classFields.ts:
```diff
@@ -664,23 +678,57 @@ namespace ts {
+                        }
+                        // If a synthetic super() call was added, add them just after it
+                        else if (needsSyntheticConstructor) {
+                            statements = addRange(
```
Why use `addRange` here? Why not just do `[statements[0], ...parameterProperties, ...statements.slice(1)]`?

rbuckton commented on src/compiler/transformers/classFields.ts:
```diff
@@ -632,10 +632,27 @@ namespace ts {
+                foundSuperStatement = initialDirectivesAndSuperCall.foundSuperStatement;
+
+                // Add existing statements before any super call
+                statements.splice(
```
I'm not a huge fan of using `splice`. It'd be better if we can rework this such that calling `splice` isn't required.

andrewbranch commented:
@typescript-bot perf test this

rbuckton commented on tests/baselines/reference/derivedClassSuperProperties.errors.txt:
```diff
@@ -0,0 +1,606 @@
+            constructor() {
+                for(a; super();) {}
+                       ~~~~~~~
+!!! error TS1345: An expression of type 'void' cannot be tested for truthiness.
```
This error is interesting because the result of `super()` is certainly not `void`. This isn't something you need to address, however.

rbuckton commented:
@JoshuaKGoldberg it looks like there are some baseline differences that were missed: https://github.com/microsoft/TypeScript/pull/29374/checks?check_run_id=1671242352#step:7:86

JoshuaKGoldberg commented:
Indeed, I rushed the merge: will tackle this weekend.

JoshuaKGoldberg commented on tests/baselines/reference/privateNameComputedPropertyName1(target=esnext).js:
```diff
@@ -41,10 +41,10 @@ new A().test();
         this.#b = 'b';
+        this.#c = 'c';
         this.#d = 'd';
+        this.#e = '';
```
😕 , this looks like an introduced bug.

JoshuaKGoldberg commented on tests/baselines/reference/privateNameFieldAssignment.js:
```diff
@@ -52,8 +52,9 @@ var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (
-        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m;
         _field.set(this, 0);
+        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m;
+        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m;
```
Introduced bug: these temporary variables are duplicated now...

JoshuaKGoldberg commented:
I think I've addressed everything _except_ the differences in emit in `derivedClassSuperProperties.js`. Will attempt to get that done soon..!

JoshuaKGoldberg commented:
Ok @rbuckton, I think I've addressed all the comments. Thanks for the thorough review! A few notes:

* I ended up not fully moving what is now `splitConstructorBodyStatementsOnSuper` into inline statements, since it has a loop and early exit condition
* I think the majority of out-of-order emits in _derivedClassSuperProperties.js_ were from not factoring in parentheses properly in what is now `getSuperCallFromStatement`
* With the little bit of extra inlining thanks to the review, I'd be curious to see if the perf tests are slightly better...

andrewbranch commented:
> With the little bit of extra inlining thanks to the review, I'd be curious to see if the perf tests are slightly better...

The perf impact is already within the margin of error—the scenarios were split about 50/50 on increasing/decreasing check time by minuscule amounts, so the real impact must be near zero, at least for the code we’re measuring.

rbuckton commented on src/compiler/transformers/es2015.ts:
```diff
@@ -1089,19 +1100,43 @@ namespace ts {
+                if (superCall) {
+                    // With a super() call, split the statements into pre-super() and 'body' (post-super())
+                    return {
+                        bodyStatements: factory.createNodeArray(originalBodyStatements.slice(i + 1)),
```
The `visitNodes` function accepts a `start` and a `count`. Rather than allocating a new array here, you could just return the offsets here and pass the requisite offsets into `visitNodes` above.

JoshuaKGoldberg replied:
Great, done for `superStatementIndex` instead of `preSuperStatements`. I don't see how to get that working with `bodyStatements` though, since it's passed directly to `factory.copyStandardPrologue`?

rbuckton replied:
For `copyStandardPrologue` you can just use `.slice` since that takes a `readonly Statement[]` rather than a `NodeArray<Statement>`, so there's no need to go through the effort of converting it into a `NodeArray`. That said, we could possibly add an offset to `copyStandardPrologue` similar to the one we have for `copyCustomPrologue`.

rbuckton commented on src/compiler/transformers/ts.ts:
```diff
@@ -1961,10 +1969,19 @@ namespace ts {
-            // Add the existing statements, skipping the initial super call.
-            addRange(statements, visitNodes(body.statements, visitor, isStatement, indexOfFirstStatement));
+            // Add remaining statements from the body, skipping the super() call if it was found
+            addRange(statements, visitNodes(body.statements, visitor, isStatement, superStatementIndex === undefined ? 0 : superStatementIndex + 1));
```
Another case where, if `findSuperStatementIndex` returns `-1` instead of `undefined`, this code could be simplified:

```suggestion
            addRange(statements, visitNodes(body.statements, visitor, isStatement, superStatementIndex + 1));
```

JoshuaKGoldberg replied:
There's something very aesthetically pleasing about how this happens to work out...

rbuckton reviewed (changes requested):
A few more small recommendations, then this looks good to merge.

rbuckton reviewed (approved):
This looks good. I'm going to have the bot run a few additional test suites and then I think we can merge.

@typescript-bot test this
@typescript-bot user test this

rbuckton commented:
@typescript-bot test this
@typescript-bot user test this

rbuckton commented:
Looks like there were failures in the user and RWC tests, I'm investigating to see if they are due to this change.

rbuckton commented:
The user test suite seems fine as far as I can tell. The build reports more changes than the bot PR indicates though.

rbuckton commented on src/compiler/transformers/es2015.ts:
```diff
@@ -980,30 +980,32 @@ namespace ts {
             // The assumption is that no prior step in the pipeline has added any prologue directives.
-            let statementOffset = 0;
-            if (!hasSynthesizedSuper) statementOffset = factory.copyStandardPrologue(constructor.body.statements, prologue, /*ensureUseStrict*/ false);
+            let statementOffset = postSuperStatementsStart;
```
There's a few problems in our RWC suite. The transpilation for initializers is happening *after* the statements that come before `super()`, but directly before `super()`, for example:

```ts
// source
declare class Base {
  constructor(a: any);
}
class Derived extends Base {
  x: boolean;
  constructor(a: any, x = false) {
    if (!a) {
      x = false;
    }
    super(a);
    this.x = x;
  }
}

// prior output (ES5)
var Derived = (function(_super) {
  function Derived(a, x) {
    if (x === void 0) { x = false; } // parameter initializer
    var _this = this;
    if (!a) {
      x = false;
    }
    _this = _super.call(this, a) || this;
    _this.x = x;
  }
})(Base);

// new output (ES5)
var Derived = (function(_super) {
  function Derived(a, x) {
    var _this = this;
    if (!a) {
      x = false;
    }
    if (x === void 0) { x = false; } // parameter initializer (moved)
    _this = _super.call(this, a) || this;
    _this.x = x;
  }
})(Base);
```

The updated output runs the parameter initializers too late. I imagine the same thing is happening for rest parameters. Can you add a test case to verify?

JoshuaKGoldberg replied:
Funny enough, `tests/baselines/reference/thisInConstructorParameter2.js` already did verify the default parameter portions; it just got buried among the other changes.

rbuckton commented:
I also found a few cases where the comments have moved:

```ts
// source
class Derived extends Base {
  x: any;
  constructor() {
    console.log();
    // comment
    super();
    this.x = null;
  }
}

// before
...
function Derived() {
  var _this = this;
  console.log();
  // comment
  _this = _super.call(this) || this;
}

// after
...
function Derived() {
  var _this = this;
  console.log();
  _this =
  // comment
  _super.call(this) || this;
}
```

rbuckton commented:
The user test suite is failing without reporting some errors. I'll have to check out this branch locally and run them.

JoshuaKGoldberg commented:
Interesting about the comments: I didn't expect to fix them in 6319af5, but the test case in e434962 shows them fixed...

rbuckton commented on tests/baselines/reference/derivedClassSuperStatementPosition.js:
```diff
@@ -73,6 +106,49 @@ var DerivedBasic = /** @class */ (function (_super) {
+        console.log();
+        _this = 
+        // c2
+        _super.call(this) || this;
```
This is still another odd comment placement. I think you need to set the comment range for the `_this = super.call(this) || this` expression to the original `super()` expression.

rbuckton replied:
The RWC tests are showing this comment behavior in a number of places as well.

rbuckton reviewed (changes requested):
One minor nitpick, other than that this is looking ready to merge. I'm going to trigger the bot to run the user and RWC tests again to very there aren't any other issues we're missing.

rbuckton commented:
@typescript-bot test this
@typescript-bot user test this

rbuckton commented:
The RWC tests look good from what I can see, with the exception of the comment emit bug that I have referenced. I think once the comment emit is addressed, this will be ready to merge.

@DanielRosenwasser - Is this a change we would consider for the 4.2 RC or will we need to wait to merge until after RC?

rbuckton commented:
Re-running user tests now that they all should be operational once again...

@typescript-bot user test this

rbuckton commented:
@typescript-bot test this

rbuckton commented:
Per our RWC tests, there's still one more case with a missing comment (from an old version of babylon.js):

**input:**
```ts
export class _Primitive extends Geometry {
    // Private 
    private _beingRegenerated: boolean;
    private _canBeRegenerated: boolean;

    constructor(id: string, scene: Scene, vertexData?: VertexData, canBeRegenerated?: boolean, mesh?: Mesh) {
        this._beingRegenerated = true;
        this._canBeRegenerated = canBeRegenerated;
        super(id, scene, vertexData, false, mesh); // updatable = false to be sure not to update vertices
        this._beingRegenerated = false;
    }
   ...
}
```

**output:**
![image](https://user-images.githubusercontent.com/3902892/105754804-5921f080-5eff-11eb-8932-73b325650552.png)

rbuckton commented:
Interesting that that old code incorrectly access `this` before `super()`.

rbuckton commented:
Ah, I found the issue. You are setting the comment range of the _ExpressionStatement_ to `superExpression`, rather than the statement in which `superExpression` is contained. The _ExpressionStatement_'s comment range ends after the trailing `;`, so the comment following the trailing `;` is lost. You need to pass the original _ExpressionStatement_ via an additional argument to `insertSuperThisCaptureThisForNode`/`insertCaptureThisForNode` to get the correct comment mapping.

JoshuaKGoldberg commented:
Ok @rbuckton! 4.2 is released _(congrats!)_ and the latest `master` is merged into this PR. I believe I've handled https://github.com/microsoft/TypeScript/pull/29374#issuecomment-767100206 in 44d7a9a252db42afdfc9194fb5ee364a86e5cfc6. Could you please review again?

JoshuaKGoldberg commented:
Ping @rbuckton - sorry to keep bothering you on this one! Now that 4.3 is in RC do you think this is ready for final review -> merging for 4.4? I don't think any of the merge conflicts introduced since January have significantly changed the logic. Thanks!

rbuckton commented on src/compiler/transformers/classFields.ts:
```diff
@@ -1315,16 +1356,26 @@ namespace ts {
          */
-        function addPropertyOrClassStaticBlockStatements(statements: Statement[], properties: readonly (PropertyDeclaration | ClassStaticBlockDeclaration)[], receiver: LeftHandSideExpression) {
-            for (const property of properties) {
+        function addPropertyOrClassStaticBlockStatements(statements: Statement[], properties: readonly (PropertyDeclaration | ClassStaticBlockDeclaration)[], receiver: LeftHandSideExpression, insertionIndex?: number) {
```
I don't see a call to this that uses `insertionIndex`, is it needed?

JoshuaKGoldberg replied:
Aha, it used to be. No longer. Removing.

rbuckton commented on tests/cases/conformance/classes/constructorDeclarations/superCalls/derivedClassParameterProperties.ts:
```diff
@@ -56,7 +56,7 @@ class Derived7 extends Base {
         this.a = 3;
         this.b = 3;
-        super(); // error
+        super();
```
Still an error (because of the `this` reference on the line above)

JoshuaKGoldberg replied:
I had removed a bunch of comments like this one since they were falling out of sync with the `.errors.txt` baselines. I'll go ahead and finish removing the rest of them from this file -- mostly `// ok`s.

JoshuaKGoldberg replied:
I had removed a bunch of comments like this one since they were falling out of sync with the `.errors.txt` baselines. I'll go ahead and finish removing the rest of them from this file -- mostly `// ok`s.

rbuckton commented on tests/cases/conformance/classes/constructorDeclarations/superCalls/derivedClassParameterProperties.ts:
```diff
@@ -79,7 +79,7 @@ class Derived9<T> extends Base2<T> {
         this.a = 3;
         this.b = 3;
-        super(); // error
+        super();
```
Still an error (because of the `this` reference on the line above)

rbuckton reviewed (changes requested):
This keeps falling off my radar, which is unfortunate because this is something I've wanted for quite a while. I have a few more comments and I'm going to kick off the various extended test suites again to make sure there's nothing else that's missing.

rbuckton commented:
@typescript-bot perf test
@typescript-bot run dt
@typescript-bot test this
@typescript-bot user test this

pkoch commented:
@rbuckton Is there anything else that @JoshuaKGoldberg can do to move this forward?

rbuckton commented:
@typescript-bot user test this inline

rbuckton commented:
> @rbuckton Is there anything else that @JoshuaKGoldberg can do to move this forward?

Nope, it looks ready to go actually, and just in time for 4.6 beta.

JoshuaKGoldberg commented:
Fantastic, thanks so much for the reviews & merge @rbuckton!

If any issues come out of this change I'm available to try to fix, if that's helpful.

NaveedAhmadHematmal commented:
The longest-living PR I have ever seen. 😳

lukescott commented:
YES!!!🎉 🎉 🎉

callumok2004 commented:
🥳🎉🎉

iJungleboy commented:
Amazing work and perseverance. Awesome!

taylus commented:
grats on the PR

upq commented:
Congrats !! Happy for you!
Might as well be the most famous PR yet ... from branding perspective I recommend you naming the PR 
Suggestion: Des Supa
